CDAP-3521 Recording lineage information for a dataset accessed by a flowlet

poornachandra opened the pull request:
Only recording lineage for datasets from flowlets in this PR to keep changes minimal. I'll add recording for streams and other program types in an upcoming PR.

Also, the following TODOs will be fixed in an upcoming PR:
1. Avoid duplicate writes of lineage for a dataset instance.
2. Add complete metadata and tags when recording lineage.

The PR looks big due to refactoring of tests and new classes for serializing lineage over HTTP.

poornachandra commented:
Build - http://builds.cask.co/browse/CDAP-DUT2810-8

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java:
```diff
@@ -0,0 +1,66 @@
+  private final LineageStore lineageStore;
+
+  @Inject
+  public BasicLineageWriter(BusinessMetadataStore businessMetadataStore, LineageStore lineageStore) {
```
package private

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriter.java:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  private String gatherMetadata(Id.Run run, Id.DatasetInstance datasetInstance) {
+    // TODO: add app metadata and tags
```
this is available, right? just that you need to create an object to represent all this info? Makes sense to do in another PR

poornachandra replied:
Yes.

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java:
```diff
@@ -0,0 +1,206 @@
+  private final AtomicReference<Id.NamespacedId> componentIdRef = new AtomicReference<>();
+
+  @Inject
+  public LineageWriterDatasetFramework(
```
package private

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/metadata/serialize/RelationProto.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Class to serialize {@link co.cask.cdap.data2.metadata.lineage.Relation}.
+ */
+public class RelationProto {
```
should these classes move to ``cdap-proto``?

poornachandra replied:
Since these classes are only used for Lineage now, let's leave it here. It should be easy to move them when required, right?

bdmogal replied:
``LineageProto`` is the output of the lineage REST API, though, right? Its ok to do it in another PR, but we should do it in 3.2?

poornachandra replied:
If all classes used as output of REST API have to go to cdap-proto then I'll move them.

bdmogal replied:
yes, that's the convention. It's ok to create a move-only PR after this one though.

poornachandra replied:
Will do.

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataSetsModules.java:
```diff
@@ -23,16 +23,22 @@
  */
 public class DataSetsModules extends RuntimeModule {
 
+  public static final String BASIC_DATASET_FRAMEWORK = "basicDatasetFramework";
```
should we call this ``delegateDatasetFramework`` to indicate that the ``LineageDatasetFramework`` will delegate to these classes?

poornachandra replied:
`basicDatasetFramework` is used by classes other than `LineageDatasetFramework`, for them delegate does not add any meaning I think.

bdmogal replied:
Ok, maybe I missed it. Will the ``DatasetFramework`` named as ``basicDatasetFramework`` be used by anything other than the ``LineageDatasetFramework``?

poornachandra replied:
Yes, it will be used by classes that `LineageDatasetFramework` depends on, such as `BusinessMetadataStore` and `LineageStore` to prevent circular dependency.

bdmogal replied:
Ok clearly I missed that. **basic** is reasonable.

bdmogal commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/Lineage.java:
```diff
@@ -27,27 +26,15 @@
  */
 public class Lineage {
   private final Set<Relation> relations;
-  private final Set<Id.Program> programs;
```
how come these two were removed? Just the fact that they can be derived from the set of relations? I think they provided a nice utility, instead of users having to compute this themselves?

poornachandra replied:
Right now they are not being used anywhere. Having them only increased the size of code, hence removed them. It should be easy to add them back if required later.

bdmogal replied:
Sounds good.

bdmogal commented:
@poornachandra the approach makes sense to me and is easy to understand. Only major comments  about moving the proto classes to ``cdap-proto``, and the removal of programs and data from the ``Lineage`` class.

poornachandra commented:
@bdmogal Thanks for the review. Addressed your comments. Can you please take another look?

Also, can I make the package-private changes in the upcoming PR?

bdmogal commented:
Ok to make changes in another PR since multiple PRs depend on each other here. LGTM.

poornachandra commented:
Build passed, merging.
